File: backend/app/ai/eligibility_engine.py
from typing import Dict, Any, List, Tuple

class EligibilityEngine:
    @staticmethod
    def evaluate(citizen_profile: Dict[str, Any], service: Dict[str, Any]) -> Tuple[bool, str, List[str]]:
        """
        Evaluates preliminary eligibility deterministically.
        Returns: (is_eligible, status_text, matching_rules)
        """
        criteria = service.get("eligibility_criteria", {})
        matching_rules = []
        violations = []

        # 1. Income Check
        max_income = criteria.get("max_income")
        user_income = citizen_profile.get("annual_income", 0)
        if max_income is not None:
            if user_income <= max_income:
                matching_rules.append(f"Annual income ₹{user_income:,} is within limit (₹{max_income:,})")
            else:
                violations.append(f"Annual income ₹{user_income:,} exceeds ₹{max_income:,}")

        # 2. Occupation Check
        allowed_occupations = criteria.get("occupations", [])
        user_occ = citizen_profile.get("occupation", "")
        if allowed_occupations:
            if user_occ.lower() in [o.lower() for o in allowed_occupations] or "All" in allowed_occupations:
                matching_rules.append(f"Occupation ({user_occ}) meets target criteria")
            else:
                violations.append(f"Occupation ({user_occ}) is not eligible (Requires: {', '.join(allowed_occupations)})")

        # 3. Age Check
        min_age = criteria.get("min_age")
        max_age = criteria.get("max_age")
        user_age = citizen_profile.get("age", 20)
        if min_age and user_age < min_age:
            violations.append(f"Age ({user_age}) is below minimum requirement ({min_age})")
        elif max_age and user_age > max_age:
            violations.append(f"Age ({user_age}) exceeds maximum limit ({max_age})")
        else:
            if min_age or max_age:
                matching_rules.append(f"Age ({user_age}) meets target range")

        # Result calculation
        if not violations:
            return True, "Eligible", matching_rules
        elif len(matching_rules) > 0:
            return False, "Partially Eligible", violations
        else:
            return False, "Not Eligible", violations

eligibility_engine = EligibilityEngine()

File: frontend/src/mock/eligibilityEngine.ts
export interface EligibilityResult {
  is_eligible: boolean;
  status_text: string;
  matching_rules: string[];
  violations: string[];
}

export class EligibilityEngine {
  static evaluate(
    citizenProfile: { annual_income?: number; occupation?: string; dob?: string; age?: number },
    service: { eligibility_criteria?: { max_income?: number; occupations?: string[]; min_age?: number; max_age?: number } }
  ): EligibilityResult {
    const criteria = service.eligibility_criteria || {};
    const matching_rules: string[] = [];
    const violations: string[] = [];

    // 1. Income Check
    const maxIncome = criteria.max_income;
    const userIncome = citizenProfile.annual_income ?? 0;
    if (maxIncome !== undefined && maxIncome !== null) {
      if (userIncome <= maxIncome) {
        matching_rules.push(`Annual income ₹${userIncome.toLocaleString("en-IN")} is within limit (₹${maxIncome.toLocaleString("en-IN")})`);
      } else {
        violations.push(`Annual income ₹${userIncome.toLocaleString("en-IN")} exceeds ₹${maxIncome.toLocaleString("en-IN")}`);
      }
    }

    // 2. Occupation Check
    const allowedOccupations = criteria.occupations || [];
    const userOcc = citizenProfile.occupation || "";
    if (allowedOccupations.length > 0) {
      const isAllowed =
        allowedOccupations.some((o) => o.toLowerCase() === userOcc.toLowerCase()) ||
        allowedOccupations.includes("All");
      if (isAllowed) {
        matching_rules.push(`Occupation (${userOcc}) meets target criteria`);
      } else {
        violations.push(`Occupation (${userOcc}) is not eligible (Requires: ${allowedOccupations.join(", ")})`);
      }
    }

    // 3. Age Check
    let userAge = citizenProfile.age;
    if (userAge === undefined && citizenProfile.dob) {
      const birthYear = new Date(citizenProfile.dob).getFullYear();
      const currentYear = new Date().getFullYear();
      userAge = currentYear - birthYear;
    }
    if (userAge === undefined) userAge = 22; // default fallback

    const minAge = criteria.min_age;
    const maxAge = criteria.max_age;
    if (minAge && userAge < minAge) {
      violations.push(`Age (${userAge}) is below minimum requirement (${minAge})`);
    } else if (maxAge && userAge > maxAge) {
      violations.push(`Age (${userAge}) exceeds maximum limit (${maxAge})`);
    } else if (minAge || maxAge) {
      matching_rules.push(`Age (${userAge}) meets target range`);
    }

    let isEligible = false;
    let statusText = "Not Eligible";

    if (violations.length === 0) {
      isEligible = true;
      statusText = "Eligible";
    } else if (matching_rules.length > 0) {
      isEligible = false;
      statusText = "Partially Eligible";
    }

    return {
      is_eligible: isEligible,
      status_text: statusText,
      matching_rules,
      violations,
    };
  }
}
